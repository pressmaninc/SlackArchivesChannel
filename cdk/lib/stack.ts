import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as events from '@aws-cdk/aws-events';
import * as eventsTargets from '@aws-cdk/aws-events-targets';
import * as path from 'path';

export class Stack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const fn = new lambda.Function(this, 'slack_archives_channel', {
      runtime: lambda.Runtime.PYTHON_3_8,
      // lambdaディレクトリ内にあるzipファイルをデプロイ対象とする
      code: lambda.Code.fromAsset(path.join(__dirname, '../../lambda/deploy_package.zip')),
      handler: 'lambda_function.lambda_handler',
      timeout: cdk.Duration.seconds(60),
      environment: {
        // SLACKのAPI TOKENはlambdaの定数に持たせる
        SLACK_API_TOKEN: this.node.tryGetContext("SLACK_API_TOKEN")
      }
    });

    // 毎週月曜、日本時刻で9時に発火する
    new events.Rule(this, "SlackArchivesChannelScheduleEvent", {
      // 注意：GMTで指定
      schedule: events.Schedule.cron({
        minute: "0",
        hour: "0",
        weekDay: "MON"
      }),
      targets: [new eventsTargets.LambdaFunction(fn)]
    });
  }
}
