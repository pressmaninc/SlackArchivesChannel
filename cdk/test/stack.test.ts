import { SynthUtils } from '@aws-cdk/assert'
import { App } from '@aws-cdk/core'

import { Stack } from '../lib/stack'

test('stack', () => {
  const app = new App(
    {
      context: {
        SLACK_API_TOKEN: 'xoxp'
      }
    }
  );

  const stack = new Stack(app, 'Stack');
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
