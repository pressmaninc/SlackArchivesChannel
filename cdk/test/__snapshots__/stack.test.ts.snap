// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stack 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters69c8a15d1328a084254aafbfa6db5ed51c7ea12753167169bcca78571b7a3c84ArtifactHash274310B6": Object {
      "Description": "Artifact hash for asset \\"69c8a15d1328a084254aafbfa6db5ed51c7ea12753167169bcca78571b7a3c84\\"",
      "Type": "String",
    },
    "AssetParameters69c8a15d1328a084254aafbfa6db5ed51c7ea12753167169bcca78571b7a3c84S3Bucket7D2DCE82": Object {
      "Description": "S3 bucket for asset \\"69c8a15d1328a084254aafbfa6db5ed51c7ea12753167169bcca78571b7a3c84\\"",
      "Type": "String",
    },
    "AssetParameters69c8a15d1328a084254aafbfa6db5ed51c7ea12753167169bcca78571b7a3c84S3VersionKeyCD49B408": Object {
      "Description": "S3 key for asset version \\"69c8a15d1328a084254aafbfa6db5ed51c7ea12753167169bcca78571b7a3c84\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "SlackArchivesChannelScheduleEventBED625EC": Object {
      "Properties": Object {
        "ScheduleExpression": "cron(0 0 ? * MON *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "slackarchiveschannelE466D073",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "slackarchiveschannelAllowEventRuleStackSlackArchivesChannelScheduleEventA48772690C45EB34": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "slackarchiveschannelE466D073",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "SlackArchivesChannelScheduleEventBED625EC",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "slackarchiveschannelE466D073": Object {
      "DependsOn": Array [
        "slackarchiveschannelServiceRole725E0307",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters69c8a15d1328a084254aafbfa6db5ed51c7ea12753167169bcca78571b7a3c84S3Bucket7D2DCE82",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters69c8a15d1328a084254aafbfa6db5ed51c7ea12753167169bcca78571b7a3c84S3VersionKeyCD49B408",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters69c8a15d1328a084254aafbfa6db5ed51c7ea12753167169bcca78571b7a3c84S3VersionKeyCD49B408",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "SLACK_API_TOKEN": "xoxp",
          },
        },
        "Handler": "lambda_function.lambda_handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "slackarchiveschannelServiceRole725E0307",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "slackarchiveschannelServiceRole725E0307": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
